pkg/standalone-utils/contracts/ProtocolFeeSplitter.sol

```diff
@@ -0,0 +1,152 @@
+    uint256 public defaultRevenueSharingFeePercentage;
+
+    // poolId => PoolSettings
+    mapping(bytes32 => RevenueShareSettings) public poolSettings;
+
+    constructor(IProtocolFeesCollector _protocolFeesCollector, address _treasury)
+        // The ProtocolFeeSplitter is a singleton, so it simply uses its own address to disambiguate action
+        // identifiers.
+        Authentication(bytes32(uint256(address(this))))
+    {
+        protocolFeesCollector = _protocolFeesCollector;
+        treasury = _treasury;
+    }
+
+    function setRevenueSharingFeePercentage(bytes32 poolId, uint256 newSwapFeePercentage)
+        external
+        override
+        authenticate
+    {
+        _require(newSwapFeePercentage >= _MIN_REVENUE_SHARING_FEE_PERCENTAGE, Errors.SPLITTER_FEE_PERCENTAGE_TOO_LOW);
+        _require(newSwapFeePercentage <= _MAX_REVENUE_SHARING_FEE_PERCENTAGE, Errors.SPLITTER_FEE_PERCENTAGE_TOO_HIGH);
+        poolSettings[poolId].revenueSharePercentageOverride = uint96(newSwapFeePercentage);
+        emit PoolRevenueShareChanged(poolId, newSwapFeePercentage);
+    }
+
+    function setDefaultRevenueSharingFeePercentage(uint256 feePercentage) external override authenticate {
+        _require(feePercentage <= _MAX_REVENUE_SHARING_FEE_PERCENTAGE, Errors.SPLITTER_FEE_PERCENTAGE_TOO_HIGH);
+        defaultRevenueSharingFeePercentage = feePercentage;
+        emit DefaultRevenueSharingFeePercentageChanged(feePercentage);
+    }
+
+    function setPoolBeneficiary(bytes32 poolId, address newBeneficiary) external override {
+        _require(msg.sender == Pool(VaultHelpers.toPoolAddress(poolId)).getOwner(), Errors.SENDER_NOT_ALLOWED);
+        poolSettings[poolId].beneficiary = newBeneficiary;
+        emit PoolBeneficiaryChanged(poolId, newBeneficiary);
+    }
+
+    function collectFees(bytes32 poolId) external override {
+        address pool = VaultHelpers.toPoolAddress(poolId);
+        IERC20 bpt = IERC20(pool);
+
+        IProtocolFeesCollector protocolFeesCollectorMemory = protocolFeesCollector;
+        uint256 feeCollectorBptBalance = bpt.balanceOf(address(protocolFeesCollectorMemory));
+        if (feeCollectorBptBalance == 0) {
+            _revert(Errors.NO_BPT_FEES_COLLECTED);
+        }
+
+        // withdrawCollectedFees requires array of tokens, and we are only transfering BPT token
+        IERC20[] memory tokens = new IERC20[](1);
+        tokens[0] = bpt;
+
+        address beneficiary = poolSettings[poolId].beneficiary;
+
+        if (beneficiary == address(0)) {
+            uint256[] memory amounts = new uint256[](1);
+            amounts[0] = feeCollectorBptBalance;
+            protocolFeesCollectorMemory.withdrawCollectedFees(tokens, amounts, treasury);
+            emit FeesCollected(poolId, beneficiary, 0, treasury, feeCollectorBptBalance);
+        } else {
+            uint256[] memory amounts = new uint256[](1);
+
+            uint256 poolFeeOverride = poolSettings[poolId].revenueSharePercentageOverride;
```

Comment:
This has a critical bug. `VaultHelpers.toPoolAddress` will simply remove the lower bits, leaving just the address, but we never check for it to be valid in any way. This means anyone can call `collectFees` with any `poolId` that resolves to the same address, and e.g. force collection in a way in which there's no beneficiary.

Similarly, if governance were to reduce a Pool's share to be below the default, the Pool owner could similarly set a benecifiary using an alternative `poolId` and sidestep the fee override that way.

There's a couple ways to solve this. One would be to check that the PoolId is legitimate in the Vault by e.g. calling `getPool` (and using this instead of `VaultHelpers`). Another would be to simply pass the Pool's address. I prefer the former, since it remains closer to the Vault's Pool model and doesn't assume too much about the Pool's (though we already assume it features `getOwner()` and is a token).